### PR TITLE
Implement Reek todo flag.

### LINKED ISSUE #723: reek --todo
Quoting from a discussion with @chastell so that we don't forget about this:

1. (Important) Introduce a `reek --todo` (or compatible) flag that
will output a todo.reek file which silences all current smells. This
would allow people to have a policy of ‘flag all new smells, let’s
go through the legacy ones one-by-one’. RuboCop’s --auto-gen-config
worked really well for me in a few projects.

2. (Not very) Verify that whitelisted smells are still there and
yell if they’re not. RuboCop does this and the benefit is that
(a) once you fix a smell you’re prompted to drop the whitelisting
and won’t reintroduce it unknowingly and (b) if you move / rename
a method you won’t have cruft entries forever in your reek config.

### PR DESCRIPTION
**Pull request by troessner**
First draft to fix #723  - just a skeleton at the moment.

I'd like to discuss this skeleton to make sure that we are all on the same page here.

@chastell @mvz what do you think?

### CODE REVIEW

File: lib/reek/cli/options.rb [hunk 1]
```diff
@@ -104,6 +96,24 @@ def set_configuration_options
...
       end
 
+      def set_generate_todo_list_options
+        parser.separator '\nGenerate a todo list:'
+        parser.on('-t', '--todo', 'Generate a todo list') do
+          self.generate_todo_list = true
+        end
+      end
+
+      def set_alternative_formatter_options
```
Comment (troessner) [inline lib/reek/cli/options.rb]: I just moved this method down since it was too far up before.

**Comment by mvz** [comment]
Looks good. How about calling the generated file `todo.reek` or something?

**Comment by troessner** [comment]
>> Looks good. How about calling the generated file todo.reek or something?

Excellent suggestion, updated it to "todo.reek".

File: features/generate_todo_list.feature [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      IrresponsibleModule:
+        exclude:
+        - Dirty
+      UncommunicativeMethodName:
+        exclude:
+        - Dirty#a
+        - Dirty#b
+      """
+
+  Scenario: Respects an configuration file
```
Comment (tuexss) [inline features/generate_todo_list.feature]: `an` -> `a`

File: features/step_definitions/reek_todo/generate_todo_list_steps.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
+Given(/^a directory 'lib' with one clean file 'clean\.rb' and one dirty file 'dirty\.rb'$/) do
+  write_file('lib/clean.rb', <<-EOS.strip_heredoc)
```
Comment (tuexss) [inline features/step_definitions/reek_todo/generate_todo_list_steps.rb]: still no squiggly heredoc?

File: lib/reek/cli/options.rb [hunk 4]
```diff
@@ -9,7 +9,7 @@ module CLI
     #
     # See {file:docs/Command-Line-Options.md} for details.
     #
-    # :reek:TooManyInstanceVariables: { max_instance_variables: 9 }
+    # :reek:TooManyInstanceVariables: { max_instance_variables: 10 }
```
Comment (tuexss) [inline lib/reek/cli/options.rb]: how could this be solved? replacing the bunch of instance variables with a struct? class variables?

Comment (troessner) [inline lib/reek/cli/options.rb]: >> replacing the bunch of instance variables with a struct?

This wouldn't fix the smell, just defer it to somewhere else. 

>>  class variables?

That would actually make it worse since class variables are evil. Not to mention that we actually have a [smell detector](https://github.com/troessner/reek/blob/master/docs/Class-Variable.md) for it :laughing: 

One solution I had in the back of my mind was to use composition, which in this case means to have several different objects encapsulating groups of cli arguments. We could e.g. group / abstract the exit codes in one object, and the format arguments in another. 
Since I find this "put all arguments into one instance annoying and slightly confusing as well I'll probably look into fixing this sooner or later (but in another PR then).

File: features/step_definitions/reek_todo/generate_todo_list_steps.rb (hunk 3 above)
Comment (troessner) [inline features/step_definitions/reek_todo/generate_todo_list_steps.rb]: No, because that would limit us to 2.3. And right now we're supporting all rubies [up from 2.0](https://github.com/troessner/reek/blob/master/.travis.yml#L8)

File: lib/reek/cli/options.rb (hunk 4 above)
Comment (mvz) [inline lib/reek/cli/options.rb]: There's an open ticket for ignoring instance variables used for memoization: #807.

Comment (troessner) [inline lib/reek/cli/options.rb]: But none of the instance variables below is used for memoization @mvz :)
In any case, we should group those options. And one of these days, I'm gonna look into that.

File: features/generate_todo_list.feature [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    And the todo file looks like this:
+      """
+      ---
+      IrresponsibleModule:
+        exclude:
+        - Dirty
+      UncommunicativeMethodName:
+        exclude:
+        - Dirty#a
+        - Dirty#b
```
Comment (chastell) [inline features/generate_todo_list.feature]: I love the specificity of the generated todo; very easy to tackle the issues one-by-one. :)

**Comment by chastell** [comment]
I think this is exactly what we need, at least for starters (but I can’t think of anything missing).

**Comment by troessner** [comment]
@mvz && @chastell ready for review!

File: lib/reek/cli/command.rb [hunk 6]
```diff
@@ -1,16 +0,0 @@
-module Reek
-  module CLI
-    #
-    # Base class for all commands
-    #
-    class Command
```
Comment (troessner) [inline lib/reek/cli/command.rb]: I removed the inheritance and replaced it with a mixin below since a) the inheritance lead to spaghetti code pretty fast (e.g. jumping back and forth between parent and kid to follow the flow) and b) was wrong in the first place since it violated the Liskov substitution principle big time.

File: lib/reek/cli/command/commandable.rb [hunk 7]
```diff
@@ -0,0 +1,35 @@
+module Reek
+  module CLI
+    module Command
+      #
+      # Base class for all commands
+      #
+      # :reek:ModuleInitialize
```
Comment (troessner) [inline lib/reek/cli/command/commandable.rb]: Justified here in my mind.

File: lib/reek/cli/command/report_command.rb [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+      #
+      # A command to collect smells from a set of sources and write them out in
+      # text report format.
+      #
+      class ReportCommand
+        include Commandable
+
+        def execute(app)
+          scan_for_smells app
+          reporter.smells? ? app.report_smells : app.report_success
```
Comment (troessner) [inline lib/reek/cli/command/report_command.rb]: I'm not really happy with the [object oriented clusterfuck pattern](https://github.com/troessner/reek/issues/775) used here but this would be something for another PR since this is the way it is right now.

File: lib/reek/cli/command/todo_list_command.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+          scan_for_smells app
+          File.write '.todo.reek', groups.to_yaml
+          puts "\n'.todo.reek' generated! You can now use "\
+                  'this as a starting point for your configuration.'
+        end
+
+        private
+
+        def groups
+          @groups ||= reporter.smells.group_by(&:smell_type).map do |type, smells|
```
Comment (troessner) [inline lib/reek/cli/command/todo_list_command.rb]: Credits to @backus for this elegant solution!

File: lib/reek/cli/options.rb [hunk 10]
```diff
@@ -55,10 +57,11 @@ def color_support?
         $stdout.tty?
       end
 
-      # :reek:TooManyStatements: { max_statements: 6 }
+      # :reek:TooManyStatements: { max_statements: 7 }
```
Comment (troessner) [inline lib/reek/cli/options.rb]: Justified as well. I believe we should overhaul our whole options handling but this is again something for a different PR.

**Comment by troessner** [comment]
Grummlllll. 
Fails under 2.0 with:

>> /Users/timo/dev/reek/lib/reek/cli/command/todo_list_command.rb:26:in `groups': undefined method `to_h' for []:Array (NoMethodError)

Seems like we have to use a not so elegant solution :(
I'm looking into it.

**Comment by troessner** [comment]
And green again for 2.0 ;)

File: lib/reek/cli/command.rb (hunk 6 above)
Comment (mvz) [inline lib/reek/cli/command.rb]: I'm not sure I follow how using a mixin resolves a), since you're jumping between module and class?

Does b) mean the Command pattern violates Liskov?

File: lib/reek/cli/command/commandable.rb (hunk 7 above)
Comment (mvz) [inline lib/reek/cli/command/commandable.rb]: Hm, i'm not so sure. Perhaps initialize belongs in a Command superclass, while the scanner code could be in a module, or perhaps even better in a class that the commands delegate to.

File: lib/reek/cli/command/commandable.rb [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+        attr_reader :options
+
+        def initialize(options)
+          @options = options
+        end
+
+        def scan_for_smells(app)
+          sources.each do |source|
+            reporter.add_examiner Examiner.new(source,
```
Comment (mvz) [inline lib/reek/cli/command/commandable.rb]: `reporter` is very confusing here since it's not defined inside this module.

File: lib/reek/cli/command/report_command.rb (hunk 8 above)
Comment (mvz) [inline lib/reek/cli/command/report_command.rb]: I resolved that in #830, so maybe merge that first :smile:?

File: lib/reek/cli/command/todo_list_command.rb [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+            [type, { 'exclude' => smells.map(&:context) }]
+          end
+        end
+
+        def reporter
+          @reporter ||= Report.report_class(:yaml).new
```
Comment (mvz) [inline lib/reek/cli/command/todo_list_command.rb]: I suppose the `:yaml` here is not really important? Any report class will do?

File: lib/reek/cli/options.rb (hunk 4 above)
Comment (mvz) [inline lib/reek/cli/options.rb]: Ah yes, I see. This needs some thinking on how this kind of problem should be handled in general.

File: lib/reek/cli/options.rb (hunk 10 above)
Comment (mvz) [inline lib/reek/cli/options.rb]: Agreed.

File: lib/reek/cli/command.rb (hunk 6 above)
Comment (troessner) [inline lib/reek/cli/command.rb]: >>  a), since you're jumping between module and class?

Yes. True. I still find it better readable as it is right now, but that might just be me.

>> Does b) mean the Command pattern violates Liskov?
 
Let me contrast the command pattern:

>> In object-oriented programming, the command pattern is a behavioral design pattern in which an object is used to encapsulate all information needed to perform an action or trigger an event at a later time.

with the substitution principle:

>> Substitutability is a principle in object-oriented programming. It states that, in a computer program, if S is a subtype of T, then objects of type T may be replaced with objects of type S (i.e., objects of type S may substitute objects of type T) without altering any of the desirable properties of that program (correctness, task performed, etc.). 

I think it depends on your point of view. From a "command pattern" point of view where I just execute "something" it doesn't. My angle was coming from object orientated thinking and the fact that those 2 commands are completely different. In fact, they couldn't be more different. 

But while writing this down I realize that i'm probably yak-shaving right now. This discussion would make more sense if we had dozens of different commands. 
I'll rewire the commands as separate classes.

File: lib/reek/cli/command/commandable.rb (hunk 7 above)
Comment (troessner) [inline lib/reek/cli/command/commandable.rb]: Agreed, see above, I'll change it.

File: lib/reek/cli/command/commandable.rb (hunk 11 above)
Comment (troessner) [inline lib/reek/cli/command/commandable.rb]: I'll change this anyway after our discussion, but out of interest, why do you find this confusing?
This is something you will encounter with the majority of all mixins, won't you?

File: lib/reek/cli/command/todo_list_command.rb (hunk 12 above)
Comment (troessner) [inline lib/reek/cli/command/todo_list_command.rb]: It's indeed not important. Actually now that I think of it this is highly confusing. 
Let me refactor this.

**Comment by troessner** [comment]
Revising and revamping this PR:

- Handling "no smells found" properly including specs and features
- Getting back to command classes
- Rebasing against the latest changes

I'll ping you @mvz when this is done - pretty sure tomorrow though ;)

**Comment by troessner** [comment]
@mvz ready for final review!

I addressed all comments and also increased test coverage, both in specs and features.

Lgtm as soon as possible!

File: README.md [hunk 13]
```diff
@@ -337,6 +337,42 @@ end
 
 This is an incredible powerful feature and further explained under [Smell Suppresion](docs/Smell-Suppression.md).
 
+### Generating a 'todo' list
+
+Integrating tools like `Reek` into an existing larger codebase can be daunting when you have to fix
```
Comment (mvz) [inline README.md]: Just a minor nitpick: Didn't we decide to not use backticks when mentioning Reek's name, and only use them for the command?

**Comment by mvz** [comment]
One final nit and it's good to go!

File: README.md (hunk 13 above)
Comment (troessner) [inline README.md]: Changed that. We then also need a follow-up PR:

>> grep -r '`Reek`' README.md docs/ | wc -l
      12

Comment (troessner) [inline README.md]: Actually I replaced all occurences of `reek` and `Reek` with Reek in our README and our docs. Boyscout rule ;)

**Comment by troessner** [comment]
Fixed the `reek` spelling. Merge? :)

**Comment by mvz** [comment]
Nice cleanup of the README!
